source/common/upstream/upstream_connection_pool_impl.cc

```diff
@@ -0,0 +1,173 @@
+    // the connection pooling attempt.
+    ++pool.available_connections;
+    dispatcher.post([accept_cb]() {
+      UpstreamConnectionEssence no_connection;
+      accept_cb(std::move(no_connection));
+    });
+
+    return;
+  }
+  ActiveClientPtr client = std::move(pool.clients.front());
+  pool.clients.pop_front();
+  auto shared_connection_essence =
+      std::make_shared<UpstreamConnectionEssence>(client->detachSockets());
+  client.reset();
+  dispatcher.post([accept_cb, shared_connection_essence]() {
+    accept_cb(std::move(*shared_connection_essence));
+  });
+}
+
+void UpstreamConnectionPoolImpl::onEvent(ActiveClient& client, Network::ConnectionEvent /*event*/) {
+  {
+    ActiveClientPtr removed;
+    Thread::LockGuard lock(mu_);
+    ConnectionPool& pool = connection_pools_[client.host_];
+    --pool.available_connections;
+    removed = client.removeFromList(pool.clients);
+    dispatcher_.deferredDelete(std::move(removed));
+  }
+  // TODO Need to call close or other methods?
+  // TODO Safe to delete active client and connection with locks held?
+}
+
+UpstreamConnectionPoolImpl::ActiveClient::ActiveClient(
+    UpstreamConnectionPoolImpl& parent, HostConstSharedPtr host,
+    Network::ClientConnectionPtr client_connection, uint64_t remaining_requests)
+    : parent_(parent), host_(host), client_connection_(std::move(client_connection)),
+      remaining_requests_(remaining_requests) {
+  client_connection_->addConnectionCallbacks(*this);
+
+  const absl::optional<std::chrono::milliseconds> idle_timeout = host_->cluster().idleTimeout();
+  if (idle_timeout) {
+    timeout_timer_ = parent_.dispatcher_.createTimer([this]() -> void { onTimeout(); });
+    timeout_timer_->enableTimer(idle_timeout.value());
+  }
+
+  client_connection_->addReadFilter(std::make_shared<CloseConnectionReadFilter>(*this));
+}
+
+UpstreamConnectionPool::UpstreamConnectionEssence
+UpstreamConnectionPoolImpl::ActiveClient::detachSockets() {
+  std::pair<Network::ConnectionSocketPtr, Network::TransportSocketPtr> detached =
+      client_connection_->detachSockets();
+  client_connection_.reset();
+  timeout_timer_.reset();
+  return {std::move(detached.first), std::move(detached.second), remaining_requests_};
+}
+
+void UpstreamConnectionPoolImpl::ActiveClient::closeClientCnnection() {
+  timeout_timer_.reset();
+  client_connection_->close(Network::ConnectionCloseType::NoFlush);
+}
+
+void UpstreamConnectionPoolImpl::ActiveClient::onTimeout() {
+  host_->cluster().stats().upstream_cx_idle_timeout_.inc();
+  closeClientCnnection();
+}
+
+Network::FilterStatus
+UpstreamConnectionPoolImpl::CloseConnectionReadFilter::onData(Buffer::Instance& /*data*/,
+                                                              bool /*end_stream*/) {
+  // TODO more testing. I think code path is triggered by
+  // test/integration/integration_test.cc
+  // IpVersions/IntegrationTest.AdminDrainDrainsListeners due to stray data.
+  active_client_.closeClientCnnection();
```

Comment:
Take the example of a HEAD request to an upstream that gets a reply that contains a response body instead of just headers.  It is possible for stray data to be detected by this reader callback.  When we hit this kind of HTTP error we should terminate the upstream connection.